typing/shape.ml

```diff
@@ -238,41 +238,42 @@ module Make_reduce(Params : sig
   val read_unit_shape : unit_name:string -> t option
   val find_shape : env -> Ident.t -> t
 end) = struct
-  let rec reduce env fuel t =
-    let reduce_if_gas =
-      if fuel > 0
-      then reduce env (fuel -1)
-      else Fun.id
-    in
-    let reduce = reduce env fuel in
-    match t.desc with
-    | Comp_unit unit_name ->
-        begin match Params.read_unit_shape ~unit_name with
-        | Some t -> reduce t
-        | None -> t
-        end
-    | App(f, arg) ->
-        app ?uid:t.uid (reduce f) ~arg:(reduce arg)
-    | Proj(str, item) ->
-        let r = proj ?uid:t.uid (reduce str) item in
-        if r = t
-        then t
-        else reduce r
-    | Abs(var, body) ->
-        { t with desc = Abs(var, reduce body) }
-    | Var id ->
-        begin try
-          let res = Params.find_shape env id in
-          if res = t then
-            raise Not_found
-          else
-            reduce_if_gas res
-        with Not_found -> { t with desc = Leaf } (* avoid loops. *)
-        end
-    | _ ->
-        t
-
-  let reduce env = reduce env Params.fuel
+  let fuel = ref (-1)
+  let rec reduce env t =
+    if !fuel < 0 then
+      t
+    else
+      match t.desc with
+      | Comp_unit unit_name ->
+          begin match Params.read_unit_shape ~unit_name with
+          | Some t -> reduce env t
+          | None -> t
+          end
+      | App(f, arg) ->
+          app ?uid:t.uid (reduce env f) ~arg:(reduce env arg)
+      | Proj(str, item) ->
+          let r = proj ?uid:t.uid (reduce env str) item in
+          if r = t
+          then t
+          else reduce env r
+      | Abs(var, body) ->
+          { t with desc = Abs(var, reduce env body) }
+      | Var id ->
+          begin try
+            let res = Params.find_shape env id in
+            if res = t then
+              raise Not_found
+            else
+              decr fuel;
+              reduce env res
+          with Not_found -> { t with desc = Leaf } (* avoid loops. *)
+          end
+      | _ ->
+          t
+
+  let reduce env t =
+    fuel := Params.fuel;
+    reduce env t
```

Comment:
I didn't notice this change when rereading the code, but the previous use of fuel wasn't incorrect: it is was limiting the depth of expansion of recursive module variables, whereas the current version is limiting the total number of expansions. Both strategies seem valid to me?

Reply:
I have the same understanding of the code as you, the reason for the change was that the previous version was looping on @gasche's example:
```ocaml
module rec A
  : sig val x : int -> int end
  = struct include B end
 and B
  : sig val x : int -> int end
  = struct include A end
```